### PR TITLE
"Sink type" that can provide backpressure

### PR DESCRIPTION
**Pull request by beerriot**
At the head of a pipe, and between its fittings, passing inputs/outputs is a blocking operation, which allows it to provide backpressure to keep downstream processes from being overwhelmed. Before this commit, there was no such behavior between the tail of the pipe and its sink. For very large Riak KV MapReduce jobs, it was common for the sink to become overwhelmed either with results or with log messages during failures.

This PR introduces blocking sink-delivery. It does this by providing the ability to specify the "type" of the sink, which indicates the method by which messages are delivered.

At this time there are two types:
- `raw`: The behavior that existed before this commit, and the default if none is specified. Messages are delivered by simply sending them to the sink process (`Sink ! Message.` in Erlang).
- `{fsm_sync, Timeout}`: Messages are delivered as synchronous `gen_fsm` events (`gen_fsm:sync_send_event(Sink, Message, Timeout)`).

The type is set in the `Options` parameter of `riak_pipe:exec/2` via the `sink_type` property: `riak_pipe:exec(Spec, [{sink_type, {fsm_sync, infinity}}|OtherOptions])`.

The spec of `riak_sink:*` has changed to facilitate error detection. Instead of returning the message it sent, it now returns the atom `ok` on success (always for `raw` sink types), or `{error, Reason}` for failure. Current `Reason`s are `timeout` if the request times out, and `sink_died` if the sink was gone before the call or vanished before replying.

New tests for this facility are in `riak_pipe:sink_type_test_/0`. A mock sink FSM for testing is included as `test/riak_pipe_test_sink_fsm.erl`.

Notes:
- Yes, synchronous delivery of each output is likely to introduce additional latency. This latency is likely more desirable than the alternative of unbounded memory consumption. There is also ongoing work (in the `bwf-pmi` branch) to deliver multiple outputs in a single message.
- Yes, synchronous log delivery is a hinderance to logging lots of messages. Again, it's likely more desirable than blowing up a sink with messages from many concurrent log producers. Additionally, the synchronicity is only encountered if the trace is actually delivered, and not if it is filtered out. Finally, the only current use of logging is in Riak KV's MapReduce, where it is used to deliver errors that cause the pipe to teardown immediately. Once one log is delivered there, the rate at which others can be delivered is not relevant.
- Synchronous delivery of the single eoi message seems unnecessary, but it makes the API more predictable.

### CODE REVIEW

File: test/riak_pipe_test_sink_fsm.erl [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+%% @doc Sink fsm for eunit tests. It accumulates all pipe results/logs
+%% until eoi, then delivers them to the caller of {@link
+%% get_results/1}.
+%%
+%% There are two states: `acc' and `wait'
+%%
+%% The FSM starts out in `acc' simply accumulating results and logs.
+%%
+%% If a `get_results' arrives during live, the FSM stores the `From',
```
Comment (joedevivo) [inline test/riak_pipe_test_sink_fsm.erl]: not sure what you mean by "live"

Comment (beerriot) [inline test/riak_pipe_test_sink_fsm.erl]: :P a missed doc update during a rewrite. should be `acc`

**Comment by joedevivo** [comment]
I like it. Especially how it's all backwards compatible. eunits pass, as to the java client tests with a version of kv that doesn't yet support this feature.

Tentative +1, the +1 will be official when riak_kv#429 gets plus oned, as that will be the integration test of the {fsm_sync, Timeout} code path. Hang in there @beerriot, we're close.

**Comment by joedevivo** [comment]
+1 official now

**Comment by beerriot** [comment]
Sorry to negate your +1, but performance testing of basho/riak_kv#429 showed the naive approach here to be too expensive. I've just added three more commits to address this.

The first commit allows workers to use `gen_fsm:send_event/2` for several sink messages, and only occasionally use `gen_fsm:sync_send_event/3`. The number of un-acked sends (`send_event/2`) that may be used before an acked send (`sync_send_event/2`) is described by a new element of the sink-type tuple, `Period`. It is specified as an integer number of un-acked messages to send, before sending one acked message. Alternatively, it can be set to the atom `infinity` which will cause all sends to be un-acked.

The second commit forces the _first_ message to be sent acked, unless the `Period` is `infinity`. This is to prevent worker restart from overwhelming the sink with non-acked messages, because the ack-counter is stored in the worker's process dictionary.

The third commit renames the sink type from `fsm_sync` to `fsm` because using `Period==infinity` is the same as what `fsm_async` might some day have been. Furthermore, the original definition of `fsm_sync` (acking every message) is equivalent to `Period==0`.

So, the sink type tuple is now `{fsm, Period::integer() | infinity, Timeout::timeout()}`.

**Comment by joedevivo** [comment]
+1 and thanks for all the :fish: 

Tested with riak+Test. looks good.

**Comment by beerriot** [comment]
Squashed to master as 807c632.

**Pull request closed without merging**